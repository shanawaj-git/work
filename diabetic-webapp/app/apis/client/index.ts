import { APIClientOptions } from 'apis/types';
import axios from 'axios';
import { addInterceptors } from './addInterceptors';
import * as environmentConfig from '../../configs/environmentConfig';

let httpClient;
// eslint-disable-next-line no-unused-vars
const configureAPIClient = (options: APIClientOptions) => {
  // TODO: intercept response and add callback support for login, auth failed
  httpClient = axios.create({
    baseURL:
      environmentConfig.getEnvironmentConfig().API_BASE_URL ||
      'http://localhost:1337',
    timeout: 30000, // 30 seconds
    // TODO: pass Authorization once login is implemented
  });
  addInterceptors(httpClient, options);
  return httpClient;
};

const getAPIClientInstance = () => {
  if (!httpClient) {
    throw new Error('API Client is not intialized');
  }
  return httpClient;
};

export { configureAPIClient, getAPIClientInstance, APIClientOptions };
