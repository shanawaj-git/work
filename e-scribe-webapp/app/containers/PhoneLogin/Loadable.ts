/**
 * Asynchronously loads the component for Splash screen
 */

import loadable from 'utils/loadable';
export default loadable(() => import('./index'));

