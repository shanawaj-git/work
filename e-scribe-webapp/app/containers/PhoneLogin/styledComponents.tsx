import styled from 'styled-components';
import { Input } from '@albathanext/design-system';

export const Container = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: flex-start;
  flex: 0 0 auto;
  flex-grow: 1;
  flex-shrink: 0;
  flex-basis: auto;
  min-height: 400px;
  flex-direction: column;
  padding: 0px 18px;
  max-width: 100%;
`;

export const Wrapper = styled.div`
  display: flex;
  padding-top: 50px;
  padding-bottom: 48px;
  justify-content: flex-start;
  align-items: fles-start;
  flex-direction: column;
`;

export const StyledInput = styled(Input)`
  &&.MuiFormControl-root {
    width: 100%;
  }

  .MuiInput-root {
    margin-bottom: 10px;
    width: 100%;

    input {
      width: 100%;
      border-color: #ed2c25;
      caret-color: #ed2c25;
      apperance: none;
    }
  }

  .MuiInputLabel-root {
    width: 100%;
  }
`;

export const ButtonContainer = styled.div`
  width: 100%;
  padding-bottom: 24px;
`;

export const ErrorText = styled.div`
  width: 100%;
  color: #ed2c25;
`;
