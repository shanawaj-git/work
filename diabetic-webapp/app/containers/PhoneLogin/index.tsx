import React, { useState, useEffect } from 'react';
import ScrollLock from 'react-scrolllock';
import { injectIntl } from 'react-intl';
import PropTypes, { any, InferProps } from 'prop-types';
import { Typography, TypographyType, Button } from '@albathanext/design-system';
import {
  sanitizePhoneNumber,
  isValidPhoneNumber,
  isNumeric,
  concatSupportPhoneNumber,
} from 'utils/phoneNumber';
import { viewPortHandler } from 'utils/viewPortHandler';
import { getReferrer } from 'utils/url';
import * as authApis from 'apis/auth';
import {
  Container,
  Wrapper,
  StyledInput,
  ButtonContainer,
  ErrorText,
  ClickHereLink,
  MpcLogo,
  PhoneLoginHeader,
  PhoneLoginDescription,
} from './styledComponents';

const PHONE_NUMBER_DIGITS_COUNT = 12;
const MOBILE_NUMBER_FORMATE = '971xxxxxxxxx';
const OTP_INPUT_SCREEN = 'otp-input';
const mpcGrayLogo = require('images/mpc_gray.svg');
export const GENERIC_ERROR_CODE = 'genericAPIError';

function PhoneLogin({ intl, history, location }: PhoneLoginTypes) {
  const referrer = getReferrer(location);
  const messages = {
    titlePrefix: intl.formatMessage({
      id: 'phoneLogin.titlePrefix',
      defaultMessage: 'Whatâ€™s your',
    }),
    titleSuffix: intl.formatMessage({
      id: 'phoneLogin.titleSuffix',
      defaultMessage: 'mobile number?',
    }),
    mobileNumber: intl.formatMessage({
      id: 'phoneLogin.mobileNumber',
      defaultMessage: 'Mobile number',
    }),
    mobileNumberInput: intl.formatMessage({
      id: 'phoneLogin.inputMobileMessage',
      defaultMessage: 'Please input the number registered with the clinic.',
    }),
    troubleLoginInMessage: intl.formatMessage({
      id: 'phoneLogin.acceptTermsAndConditions',
      defaultMessage: 'Having trouble logging in? ',
    }),
    clickHere: intl.formatMessage({
      id: 'phoneLogin.clickHere',
      defaultMessage: 'Click here ',
    }),
    next: intl.formatMessage({
      id: 'phoneLogin.next',
      defaultMessage: 'Continue',
    }),
    [authApis.AuthApiErrors.UnknownUser]: intl.formatMessage({
      id: 'phoneLogin.next',
      defaultMessage: 'The mobile number is not registered with us.',
    }),
    [GENERIC_ERROR_CODE]: intl.formatMessage({
      id: 'phoneLogin.genericAPIError',
      defaultMessage: 'Sorry, something went wrong',
    }),
  };

  // TODO - Should retrieve it from CMS later
  const supportPhoneNumber = concatSupportPhoneNumber('9524389675346');

  const [phoneNumber, setPhoneNumber] = useState('');

  const [errorCode, setErrorCode] = useState('');

  useEffect(() => {
    window.visualViewport.addEventListener('resize', viewPortHandler('button'));

    return () => {
      window.visualViewport.removeEventListener(
        'resize',
        viewPortHandler('button'),
      );
    };
  }, [viewPortHandler]);

  const onChange = (event: any): void => {
    const { value } = event.target;

    const sanitizedPhoneNumber = sanitizePhoneNumber(value);
    if (shouldUpdateInputField(sanitizedPhoneNumber)) {
      setPhoneNumber(sanitizedPhoneNumber);
    }
    setErrorCode(null);
  };

  const shouldUpdateInputField = (value: string): boolean => {
    if (
      value.length <= PHONE_NUMBER_DIGITS_COUNT &&
      (isNumeric(value) || value === '')
    ) {
      return true;
    }

    return false;
  };

  const getErrorMessage = code =>
    messages[code] || messages[GENERIC_ERROR_CODE];

  const onNextClick = async () => {
    setErrorCode(null);
    try {
      await authApis.generateOTP(phoneNumber);
      history.push({
        pathname: OTP_INPUT_SCREEN,
        state: { phoneNumber, referrer },
      });
    } catch (err) {
      setErrorCode(err.message);
    }
  };

  return (
    <ScrollLock>
      <Container>
        <MpcLogo src={mpcGrayLogo} />
        <div>
          <Wrapper>
            <PhoneLoginHeader>
              <Typography typographyType={TypographyType.TITLE_1}>
                {messages.titlePrefix}
              </Typography>
              <Typography typographyType={TypographyType.LARGE_TITLE}>
                {messages.titleSuffix}
              </Typography>
            </PhoneLoginHeader>
            <PhoneLoginDescription typographyType={TypographyType.SUB_HEAD}>
              {messages.mobileNumberInput}
            </PhoneLoginDescription>
          </Wrapper>

          <StyledInput
            label={messages.mobileNumber}
            onChange={onChange}
            value={phoneNumber}
            type="tel"
            placeholder={MOBILE_NUMBER_FORMATE}
          />

          <Typography typographyType={TypographyType.CAPTION}>
            {messages.troubleLoginInMessage}
          </Typography>

          <ClickHereLink
            typographyType={TypographyType.CAPTION}
            data-testid="clickHere"
          >
            <a id="contactSupport" href={supportPhoneNumber}>
              {messages.clickHere}
            </a>
          </ClickHereLink>
        </div>

        <ButtonContainer>
          <Button
            data-testid="login-btn"
            variant="contained"
            size="large"
            style={{ width: '100%' }}
            disabled={!isValidPhoneNumber(phoneNumber)}
            onClick={onNextClick}
          >
            <>{messages.next}</>
          </Button>
          {errorCode && (
            <ErrorText data-testid="error-text">
              <Typography typographyType={TypographyType.FOOT_NOTE}>
                {getErrorMessage(errorCode)}
              </Typography>
            </ErrorText>
          )}
        </ButtonContainer>
      </Container>
    </ScrollLock>
  );
}

const PhoneLoginPropTypes = {
  intl: any,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }),
  location: PropTypes.shape({
    state: PropTypes.shape({
      referrer: PropTypes.string.isRequired,
    }),
  }),
};

type PhoneLoginTypes = InferProps<typeof PhoneLoginPropTypes>;

PhoneLogin.propTypes = PhoneLoginPropTypes;

export default injectIntl(PhoneLogin);
