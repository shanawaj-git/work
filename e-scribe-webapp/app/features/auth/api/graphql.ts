import { gql } from '@apollo/client';

export type GenerateOTPInput = {
  phoneNumber: string;
};

export type GenerateOTPOutput = {
  generateOTP: {
    success: boolean;
    error: {
      code: AuthErrorCode;
      message?: string;
      resendWaitPeriodMillis?: number;
    };
  };
};

export type ValidateOTPInput = {
  phoneNumber: string;
  otp: string,
};

export type ValidateOTPOutput = {
  validateOTP: {
    success: boolean;
    error: {
      code: AuthErrorCode;
      message?: string;
    };
  };
};

export const AuthMutations = {
  GenerateOTP: gql`
    mutation GenerateAuthOTP($input: GenerateOTPInput!) {
      generateOTP(input: $input) {
        success
        error {
          code
          message
        }
      }
    }
  `,
  ValidateOTP: gql`
  mutation ValidateAuthOTP($input: ValidateOTPInput!) {
    validateOTP(input: $input) {
      success
      error {
        code
        message
      }
    }
  }
`,
};

export enum AuthErrorCode {
  AwaitingResendDelay = 'AUTH_ERR_001',
  MaxAttemptsExceeded = 'AUTH_ERR_002',
  InvalidCredentials = 'AUTH_ERR_101',
  ExpiredOTP = 'AUTH_ERR_102',
};

export const OPERATION_VALIDATE_AUTH_OTP = 'ValidateAuthOTP';
