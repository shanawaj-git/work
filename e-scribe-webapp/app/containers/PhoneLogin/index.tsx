import React, { useState, useEffect } from 'react';
import ScrollLock from 'react-scrolllock';
import { useMutation } from '@apollo/client';
import { injectIntl } from 'react-intl';
import PropTypes, { any, InferProps } from 'prop-types';
import {
  Typography,
  TypographyType,
  Button,
  SwipeableDrawer,
} from '@albathanext/design-system';
import {
  sanitizePhoneNumber,
  isValidPhoneNumber,
  isNumeric,
} from 'utils/phoneNumber';
import SubModule from 'utils/constants/env';
import { viewPortHandler } from 'utils/viewPortHandler';
import { AuthErrorCode, AuthMutations } from 'features/auth/api/graphql';
import {
  Container,
  Wrapper,
  StyledInput,
  ButtonContainer,
  ErrorText,
} from './styledComponents';

import { getReferrer } from 'utils/url';

const PHONE_NUMBER_DIGITS_COUNT = 12;
const MOBILE_NUMBER_FORMATE = '971xxxxxxxxx';
const OTP_INPUT_SCREEN = 'otp-input';
export const GENERIC_ERROR_CODE = 'genericAPIError';

function PhoneLogin({ intl, history, location }: PhoneLoginTypes) {
  const referrer = getReferrer(location);
  const messages = {
    title: intl.formatMessage({
      id: 'phoneLogin.title',
      defaultMessage: 'Whatâ€™s your mobile number?',
    }),
    mobileNumber: intl.formatMessage({
      id: 'phoneLogin.mobileNumber',
      defaultMessage: 'Mobile number',
    }),
    acceptTermsAndConditions: intl.formatMessage({
      id: 'phoneLogin.acceptTermsAndConditions',
      defaultMessage: 'By tapping next, you agree to the ',
    }),
    termsAndConditions: intl.formatMessage({
      id: 'phoneLogin.termsAndConditions',
      defaultMessage: 'Terms & Conditions',
    }),
    next: intl.formatMessage({
      id: 'phoneLogin.next',
      defaultMessage: 'Next',
    }),
    // TODO: The UX in cases of error and error messages are to be defined
    [AuthErrorCode.AwaitingResendDelay]: intl.formatMessage({
      id: 'phoneLogin.error.awaitingResendDelay',
      defaultMessage: 'Please try after some time',
    }),
    [AuthErrorCode.MaxAttemptsExceeded]: intl.formatMessage({
      id: 'phoneLogin.error.maxAttemptsExceeded',
      defaultMessage: 'Maximum attempts exceeded',
    }),
    [GENERIC_ERROR_CODE]: intl.formatMessage({
      id: 'phoneLogin.genericAPIError',
      defaultMessage: 'Sorry, something went wrong',
    }),
  };

  const [phoneNumber, setPhoneNumber] = useState('');
  const [
    isTermsAndConditionsPopUpOpen,
    setIsTermsAndConditionsPopUpOpen,
  ] = useState(false);

  const [errorCode, setErrorCode] = useState('');

  useEffect(() => {
    window.visualViewport.addEventListener('resize', viewPortHandler('button'));

    return () => {
      window.visualViewport.removeEventListener(
        'resize',
        viewPortHandler('button'),
      );
    };
  }, [viewPortHandler]);

  const onChange = (event: any): void => {
    const { value } = event.target;

    const sanitizedPhoneNumber = sanitizePhoneNumber(value);
    if (shouldUpdateInputField(sanitizedPhoneNumber)) {
      setPhoneNumber(sanitizedPhoneNumber);
    }
    setErrorCode(null);
  };

  const handleTermsAndConditionsClick = () => {
    setIsTermsAndConditionsPopUpOpen(true);
  };

  const shouldUpdateInputField = (value: string): boolean => {
    if (
      value.length <= PHONE_NUMBER_DIGITS_COUNT &&
      (isNumeric(value) || value === '')
    ) {
      return true;
    }

    return false;
  };

  const handleCloseTermsAndConditions = () => {
    setIsTermsAndConditionsPopUpOpen(false);
  };

  // TODO: to be replaced by content coming from CMS
  const renderTermsAndConditions = () => (
    <>
      <div>Terms and conditions:</div>
      <br />
      {Array.from(Array(20).keys()).map(index => (
        <div key={index}>
          <div>you agree to our terms and conditions</div>
          <br />
        </div>
      ))}
    </>
  );

  const [generateOTP, { loading }] = useMutation(AuthMutations.GenerateOTP, {
    context: { module: SubModule.AUTHENTICATION },
    variables: {
      input: {
        phoneNumber,
      },
    },
    onCompleted: data => {
      if (data.generateOTP.success) {
        history.push({
          pathname: OTP_INPUT_SCREEN,
          state: { phoneNumber, referrer },
        });
      } else {
        setErrorCode(data.generateOTP.error.code || GENERIC_ERROR_CODE);
      }
    },
    onError: () => {
      setErrorCode(GENERIC_ERROR_CODE);
    },
  });

  const getErrorMessage = code =>
    messages[code] || messages[GENERIC_ERROR_CODE];

  const onNextClick = () => {
    setErrorCode(null);
    generateOTP();
  };

  return (
    <ScrollLock>
      <Container>
        <div>
          <Wrapper>
            <Typography typographyType={TypographyType.LARGE_TITLE}>
              {messages.title}
            </Typography>
          </Wrapper>

          <StyledInput
            label={messages.mobileNumber}
            onChange={onChange}
            value={phoneNumber}
            type="tel"
            placeholder={MOBILE_NUMBER_FORMATE}
          />

          <Typography typographyType={TypographyType.CAPTION}>
            {messages.acceptTermsAndConditions}
          </Typography>

          <Typography
            onClick={() => handleTermsAndConditionsClick()}
            color="red"
            style={{ textDecoration: 'underline' }}
            typographyType={TypographyType.CAPTION}
            data-testid="termsAndConditionsLink"
          >
            {messages.termsAndConditions}
          </Typography>
        </div>
        <SwipeableDrawer
          isOpen={isTermsAndConditionsPopUpOpen}
          onClose={handleCloseTermsAndConditions}
          onOpen={() => {}}
          hasRoundedCorners
        >
          <div>{renderTermsAndConditions()}</div>
        </SwipeableDrawer>

        <ButtonContainer>
          <Button
            data-testid="login-btn"
            variant="contained"
            size="large"
            style={{ width: '100%' }}
            disabled={!isValidPhoneNumber(phoneNumber) || loading}
            onClick={onNextClick}
          >
            <>{messages.next}</>
          </Button>
          {errorCode && (
            <ErrorText data-testid="error-text">
              <Typography typographyType={TypographyType.FOOT_NOTE}>
                {getErrorMessage(errorCode)}
              </Typography>
            </ErrorText>
          )}
        </ButtonContainer>
      </Container>
    </ScrollLock>
  );
}

const PhoneLoginPropTypes = {
  intl: any,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }),
  location: PropTypes.shape({
    state: PropTypes.shape({
      referrer: PropTypes.string.isRequired,
    }),
  }),
};

type PhoneLoginTypes = InferProps<typeof PhoneLoginPropTypes>;

PhoneLogin.propTypes = PhoneLoginPropTypes;

export default injectIntl(PhoneLogin);
