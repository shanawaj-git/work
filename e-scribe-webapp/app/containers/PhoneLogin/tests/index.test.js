import React from 'react';
import { unmountComponentAtNode } from 'react-dom';
import { render, fireEvent } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { act } from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';
import { AuthErrorCode, AuthMutations } from 'features/auth/api/graphql';
import PhoneLogin from '../index';

let container = null;

const referrer = '/prescriptions/123456798';
const mockLocation = {
  state: {
    referrer,
  },
};

const WrappedPhoneLogin = () => (
  <IntlProvider locale="en">
    <MockedProvider>
      <PhoneLogin />
    </MockedProvider>
  </IntlProvider>
);

describe('<PhoneLogin />', () => {
  const phoneNumber = '971526234727';

  // eslint-disable-next-line no-global-assign
  window = Object.assign(window, {
    visualViewport: {
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
    },
  });

  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });
  afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
    jest.restoreAllMocks();
  });

  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(<WrappedPhoneLogin />);
    expect(firstChild).toMatchSnapshot();
  });

  it('should add "971" before local numbers with leading zeros', () => {
    act(() => {
      render(<WrappedPhoneLogin />, container);
    });

    // get a hold of the input element, and trigger some clicks on it
    const input = document.querySelector('.MuiInputBase-input');

    fireEvent.change(input, { target: { value: '0526234727' } });

    expect(input.value).toEqual('971526234727');
  });

  it('should add "971" before local numbers without leading zero', () => {
    act(() => {
      render(<WrappedPhoneLogin />, container);
    });

    // get a hold of the input element, and trigger some clicks on it
    const input = document.querySelector('.MuiInputBase-input');

    fireEvent.change(input, { target: { value: '526234727' } });

    expect(input.value).toEqual('971526234727');
  });

  it('should not accept non numeric values', () => {
    act(() => {
      render(<WrappedPhoneLogin />, container);
    });

    // get a hold of the input element, and trigger some clicks on it
    const input = document.querySelector('.MuiInputBase-input');

    fireEvent.change(input, { target: { value: 'xxxx$' } });

    expect(input.value).toEqual('');
  });

  it('should not accept more than 12 chars', () => {
    act(() => {
      render(<WrappedPhoneLogin />, container);
    });

    // get a hold of the input element, and trigger some clicks on it
    const input = document.querySelector('.MuiInputBase-input');

    fireEvent.change(input, { target: { value: '1234567891011213213' } });

    expect(input.value).toEqual('');
  });

  it('should open terms and conditions when clicking on the T&Cs link', () => {
    act(() => {
      render(<WrappedPhoneLogin />, container);
    });

    // get a hold of the input element, and trigger some clicks on it
    const termsAndConditionsLink = document.querySelector(
      '[data-testid="termsAndConditionsLink"]',
    );

    const overlay = document.querySelector('.MuiDrawer-modal');

    expect(overlay.classList.contains('MuiModal-hidden')).toEqual(true);

    act(() => {
      termsAndConditionsLink.dispatchEvent(
        new MouseEvent('click', { bubbles: true }),
      );
    });

    expect(overlay.classList.contains('MuiModal-hidden')).toEqual(false);

    const backdrop = document.querySelector('.MuiBackdrop-root');

    // simulate clicking on the overlay to dismiss it
    act(() => {
      backdrop.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    expect(overlay.getAttribute('aria-hidden')).toEqual('true');
  });

  it('should redirect to otp-input screen once OTP generation is successful', async () => {
    const mockHistory = {
      push: jest.fn(),
    };

    const otpInput = { phoneNumber };
    const mocks = [
      {
        request: {
          query: AuthMutations.GenerateOTP,
          variables: { input: otpInput },
        },
        result: { data: { generateOTP: { success: true, error: null } } },
      },
    ];

    act(() => {
      render(
        <IntlProvider locale="en">
          <MockedProvider addTypename={false} mocks={mocks}>
            <PhoneLogin history={mockHistory} location={mockLocation} />
          </MockedProvider>
        </IntlProvider>,

        container,
      );
    });

    const input = document.querySelector('.MuiInputBase-input');

    fireEvent.change(input, { target: { value: phoneNumber } });

    const button = document.querySelector('[data-testid="login-btn"]');

    fireEvent.click(
      button,

      new MouseEvent('click', {
        bubbles: true,

        cancelable: true,
      }),
    );
    await new Promise(resolve => setTimeout(resolve, 0)); // wait for response
    expect(mockHistory.push).toBeCalledWith({
      pathname: 'otp-input',
      state: { phoneNumber, referrer: '/prescriptions/123456798' },
    });
  });

  it('should display specific error once OTP generation has failed with specific reason', async () => {
    const mockHistory = {
      push: jest.fn(),
    };

    const otpInput = { phoneNumber };
    const mocks = [
      {
        request: {
          query: AuthMutations.GenerateOTP,
          variables: { input: otpInput },
        },
        result: {
          data: {
            generateOTP: {
              success: false,
              error: { code: AuthErrorCode.AwaitingResendDelay },
            },
          },
        },
      },
    ];

    const errorMsgCode = 'phoneLogin.error.awaitingResendDelay';

    const messages = {
      [errorMsgCode]: 'Some error message',
    };

    act(() => {
      render(
        <IntlProvider locale="en" messages={messages}>
          <MockedProvider addTypename={false} mocks={mocks}>
            <PhoneLogin history={mockHistory} />
          </MockedProvider>
        </IntlProvider>,

        container,
      );
    });

    const input = document.querySelector('.MuiInputBase-input');

    fireEvent.change(input, { target: { value: phoneNumber } });

    const button = document.querySelector('[data-testid="login-btn"]');

    fireEvent.click(
      button,

      new MouseEvent('click', {
        bubbles: true,

        cancelable: true,
      }),
    );
    await new Promise(resolve => setTimeout(resolve, 0)); // wait for response
    expect(
      document.querySelector('[data-testid="error-text"] > p').innerHTML,
    ).toEqual(expect.stringContaining(messages[errorMsgCode]));
  });

  it('should display generic error code once OTP generation has failed with unknown reason', async () => {
    const mockHistory = {
      push: jest.fn(),
    };

    const otpInput = { phoneNumber };
    const mocks = [
      {
        request: {
          query: AuthMutations.GenerateOTP,
          variables: { input: otpInput },
        },
        result: {
          data: {
            generateOTP: {
              success: false,
              error: { code: 'auth.err.unknownReason' },
            },
          },
        },
      },
    ];

    const errorMsgCode = 'phoneLogin.genericAPIError';

    const messages = {
      [errorMsgCode]: 'Sample generic error message',
    };

    act(() => {
      render(
        <IntlProvider locale="en" messages={messages}>
          <MockedProvider addTypename={false} mocks={mocks}>
            <PhoneLogin history={mockHistory} />
          </MockedProvider>
        </IntlProvider>,

        container,
      );
    });

    const input = document.querySelector('.MuiInputBase-input');

    fireEvent.change(input, { target: { value: phoneNumber } });

    const button = document.querySelector('[data-testid="login-btn"]');

    fireEvent.click(
      button,

      new MouseEvent('click', {
        bubbles: true,

        cancelable: true,
      }),
    );
    await new Promise(resolve => setTimeout(resolve, 0)); // wait for response
    expect(
      document.querySelector('[data-testid="error-text"] > p').innerHTML,
    ).toEqual(expect.stringContaining(messages[errorMsgCode]));
  });

  it('should display generic error code once OTP generation has failed with missing error code', async () => {
    const mockHistory = {
      push: jest.fn(),
    };

    const otpInput = { phoneNumber };
    const mocks = [
      {
        request: {
          query: AuthMutations.GenerateOTP,
          variables: { input: otpInput },
        },
        result: {
          data: {
            generateOTP: {
              success: false,
              error: {},
            },
          },
        },
      },
    ];

    const errorMsgCode = 'phoneLogin.genericAPIError';

    const messages = {
      [errorMsgCode]: 'Sample generic error message',
    };

    act(() => {
      render(
        <IntlProvider locale="en" messages={messages}>
          <MockedProvider addTypename={false} mocks={mocks}>
            <PhoneLogin history={mockHistory} />
          </MockedProvider>
        </IntlProvider>,

        container,
      );
    });

    const input = document.querySelector('.MuiInputBase-input');

    fireEvent.change(input, { target: { value: phoneNumber } });

    const button = document.querySelector('[data-testid="login-btn"]');

    fireEvent.click(
      button,

      new MouseEvent('click', {
        bubbles: true,

        cancelable: true,
      }),
    );
    await new Promise(resolve => setTimeout(resolve, 0)); // wait for response
    expect(
      document.querySelector('[data-testid="error-text"] > p').innerHTML,
    ).toEqual(expect.stringContaining(messages[errorMsgCode]));
  });

  it('should display generic error once OTP generation has failed for unknown reason', async () => {
    const mockHistory = {
      push: jest.fn(),
    };

    const otpInput = { phoneNumber };
    const mocks = [
      {
        request: {
          query: AuthMutations.GenerateOTP,
          variables: { input: otpInput },
        },
        error: new Error('An error occurred'),
      },
    ];

    const errorMsgCode = 'phoneLogin.genericAPIError';

    const messages = {
      [errorMsgCode]: 'Sample generic error message',
    };

    act(() => {
      render(
        <IntlProvider locale="en" messages={messages}>
          <MockedProvider addTypename={false} mocks={mocks}>
            <PhoneLogin history={mockHistory} />
          </MockedProvider>
        </IntlProvider>,

        container,
      );
    });

    const input = document.querySelector('.MuiInputBase-input');

    fireEvent.change(input, { target: { value: phoneNumber } });

    const button = document.querySelector('[data-testid="login-btn"]');

    fireEvent.click(
      button,

      new MouseEvent('click', {
        bubbles: true,

        cancelable: true,
      }),
    );
    await new Promise(resolve => setTimeout(resolve, 0)); // wait for response
    expect(
      document.querySelector('[data-testid="error-text"] > p').innerHTML,
    ).toEqual(expect.stringContaining(messages[errorMsgCode]));
  });

  it('should display generic error code once OTP generation has failed with unknown reason', async () => {
    const mockHistory = {
      push: jest.fn(),
    };

    const otpInput = { phoneNumber };
    const mocks = [
      {
        request: {
          query: AuthMutations.GenerateOTP,
          variables: { input: otpInput },
        },
        result: {
          data: {
            generateOTP: {
              success: false,
              error: { code: 'auth.err.unknownReason' },
            },
          },
        },
      },
    ];

    const errorMsgCode = 'phoneLogin.genericAPIError';

    const messages = {
      [errorMsgCode]: 'Sample generic error message',
    };

    act(() => {
      render(
        <IntlProvider locale="en" messages={messages}>
          <MockedProvider addTypename={false} mocks={mocks}>
            <PhoneLogin history={mockHistory} />
          </MockedProvider>
        </IntlProvider>,

        container,
      );
    });

    const input = document.querySelector('.MuiInputBase-input');

    fireEvent.change(input, { target: { value: phoneNumber } });

    const button = document.querySelector('[data-testid="login-btn"]');

    fireEvent.click(
      button,

      new MouseEvent('click', {
        bubbles: true,

        cancelable: true,
      }),
    );
    await new Promise(resolve => setTimeout(resolve, 0)); // wait for response
    expect(
      document.querySelector('[data-testid="error-text"] > p').innerHTML,
    ).toEqual(expect.stringContaining(messages[errorMsgCode]));
  });

  it('should display generic error code once OTP generation has failed with missing error code', async () => {
    const mockHistory = {
      push: jest.fn(),
    };

    const otpInput = { phoneNumber };
    const mocks = [
      {
        request: {
          query: AuthMutations.GenerateOTP,
          variables: { input: otpInput },
        },
        result: {
          data: {
            generateOTP: {
              success: false,
              error: {},
            },
          },
        },
      },
    ];

    const errorMsgCode = 'phoneLogin.genericAPIError';

    const messages = {
      [errorMsgCode]: 'Sample generic error message',
    };

    act(() => {
      render(
        <IntlProvider locale="en" messages={messages}>
          <MockedProvider addTypename={false} mocks={mocks}>
            <PhoneLogin history={mockHistory} />
          </MockedProvider>
        </IntlProvider>,

        container,
      );
    });

    const input = document.querySelector('.MuiInputBase-input');

    fireEvent.change(input, { target: { value: phoneNumber } });

    const button = document.querySelector('[data-testid="login-btn"]');

    fireEvent.click(
      button,

      new MouseEvent('click', {
        bubbles: true,

        cancelable: true,
      }),
    );
    await new Promise(resolve => setTimeout(resolve, 0)); // wait for response
    expect(
      document.querySelector('[data-testid="error-text"] > p').innerHTML,
    ).toEqual(expect.stringContaining(messages[errorMsgCode]));
  });

  it('should display generic error once OTP generation has failed for unknown reason', async () => {
    const mockHistory = {
      push: jest.fn(),
    };

    const otpInput = { phoneNumber };
    const mocks = [
      {
        request: {
          query: AuthMutations.GenerateOTP,
          variables: { input: otpInput },
        },
        error: new Error('An error occurred'),
      },
    ];

    const errorMsgCode = 'phoneLogin.genericAPIError';

    const messages = {
      [errorMsgCode]: 'Sample generic error message',
    };

    act(() => {
      render(
        <IntlProvider locale="en" messages={messages}>
          <MockedProvider addTypename={false} mocks={mocks}>
            <PhoneLogin history={mockHistory} />
          </MockedProvider>
        </IntlProvider>,

        container,
      );
    });

    const input = document.querySelector('.MuiInputBase-input');

    fireEvent.change(input, { target: { value: phoneNumber } });

    const button = document.querySelector('[data-testid="login-btn"]');

    fireEvent.click(
      button,

      new MouseEvent('click', {
        bubbles: true,

        cancelable: true,
      }),
    );
    await new Promise(resolve => setTimeout(resolve, 0)); // wait for response
    expect(
      document.querySelector('[data-testid="error-text"] > p').innerHTML,
    ).toEqual(expect.stringContaining(messages[errorMsgCode]));
  });
});
