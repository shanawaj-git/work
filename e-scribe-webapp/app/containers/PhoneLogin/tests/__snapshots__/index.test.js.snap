// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhoneLogin /> should render and match the snapshot 1`] = `
<div
  class="sc-bdVaJa dWYKRv"
>
  <div>
    <div
      class="sc-bwzfXH dygumy"
    >
      <h1
        class="MuiTypography-root MuiTypography-h1 css-qu1isr-MuiTypography-root"
      >
        Whatâ€™s your mobile number?
      </h1>
    </div>
    <div
      class="root MuiFormControl-root styledComponents__FormControl-lwe1h6-0 btsVJU sc-htpNat fanNAL css-1nrlq1o-MuiFormControl-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1fwwy-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for="mui-1"
          id="mui-1-label"
        >
          Mobile number
        </label>
        <div
          class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1480iag-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-invalid="false"
            autocomplete=""
            class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id="mui-1"
            name=""
            placeholder="971xxxxxxxxx"
            type="tel"
            value=""
          />
        </div>
      </div>
      
    </div>
    <span
      class="MuiTypography-root MuiTypography-caption css-1393mlh-MuiTypography-root"
    >
      By tapping next, you agree to the 
    </span>
    <span
      class="MuiTypography-root MuiTypography-caption css-115ym0p-MuiTypography-root"
      data-testid="termsAndConditionsLink"
      style="text-decoration: underline;"
    >
      Terms & Conditions
    </span>
  </div>
  <div
    class="PrivateSwipeArea-root css-17j2vre"
    style="height: 30px;"
  />
  <div
    class="sc-bxivhb kzQJYE"
  >
    <button
      class="root StyledButton-sc-1apep4b-0 hPQGaD MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root Mui-disabled  css-sghohy-MuiButtonBase-root-MuiButton-root"
      data-testid="login-btn"
      disabled=""
      style="width: 100%;"
      tabindex="-1"
      type="button"
    >
      <div
        style="visibility: visible;"
      >
        Next
      </div>
    </button>
    
  </div>
</div>
`;
